File: src/test/java/org/elasticsearch/plugin/readonlyrest/utils/LdapContainer.java
package org.elasticsearch.plugin.readonlyrest.utils;

import com.google.common.collect.Lists;
import com.unboundid.ldap.sdk.AddRequest;
import com.unboundid.ldap.sdk.BindResult;
import com.unboundid.ldap.sdk.Entry;
import com.unboundid.ldap.sdk.LDAPConnection;
import com.unboundid.ldap.sdk.LDAPConnectionOptions;
import com.unboundid.ldap.sdk.ResultCode;
import com.unboundid.ldif.LDIFReader;
import org.elasticsearch.plugin.readonlyrest.ldap.UnboundidLdapClient;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.wait.WaitStrategy;
import org.testcontainers.images.builder.ImageFromDockerfile;

import java.io.File;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.List;
import java.util.Optional;

public class LdapContainer extends GenericContainer<LdapContainer> {

    private static int LDAP_PORT = 389;
    private static Duration LDAP_CONNECT_TIMEOUT = Duration.ofSeconds(5);
    private static Duration CONTAINER_STARTUP_TIMEOUT = Duration.ofSeconds(60);
    private static String LDAP_DOMAIN = "example.com";
    private static String LDAP_ORGANISATION = "Example";
    private static String LDAP_ADMIN = "admin";
    private static String LDAP_ADMIN_PASSWORD = "password";
    private static Duration WAIT_BETWEEN_RETRIES = Duration.ofSeconds(1);

    private LdapContainer(ImageFromDockerfile imageFromDockerfile) {
        super(imageFromDockerfile);
    }

    public String getLdapHost() {
        return this.getContainerIpAddress();
    }

    public Integer getLdapPort() {
        return this.getMappedPort(LDAP_PORT);
    }

    public UnboundidLdapClient.BindDnPassword getBindDNAndPassword() {
        List<String> dnParts = Lists.newArrayList(LDAP_DOMAIN.split("\\."));
        if (dnParts.isEmpty()) throw new IllegalArgumentException("Wrong domain defined " + LDAP_DOMAIN);
        String dnString = dnParts.stream().map(part -> "dc=" + part).reduce("", (s, s2) -> s + "," + s2);
        String bindDN = String.format("cn=%s%s", LDAP_ADMIN, dnString);
        return new UnboundidLdapClient.BindDnPassword(bindDN, LDAP_ADMIN_PASSWORD);
    }

    public static LdapContainer create(String resourceFile) {
        Path initialDataLdifPath;
        try {
            initialDataLdifPath = Paths.get(LdapContainer.class.getResource(resourceFile).toURI());
        } catch (URISyntaxException e) {
            e.printStackTrace();
            throw new IllegalArgumentException(e);
        }
        LdapContainer container = new LdapContainer(
                new ImageFromDockerfile()
                        .withDockerfileFromBuilder(builder -> builder
                                .from("osixia/openldap:1.1.7")
                                .env("LDAP_ORGANISATION", LDAP_ORGANISATION)
                                .env("LDAP_DOMAIN", LDAP_DOMAIN)
                                .env("LDAP_ADMIN_PASSWORD", LDAP_ADMIN_PASSWORD)
                                .build()));
        return container
                .withExposedPorts(LDAP_PORT)
                .waitingFor(
                        container.ldapWaitStrategy(initialDataLdifPath.toFile())
                                .withStartupTimeout(CONTAINER_STARTUP_TIMEOUT)
                );
    }

    private WaitStrategy ldapWaitStrategy(File initialDataLdif) {
        return new GenericContainer.AbstractWaitStrategy() {

            @Override
            protected void waitUntilReady() {
                Optional<LDAPConnection> connection = tryConnect(getLdapHost(), getLdapPort());
                if (connection.isPresent()) {
                    try {
                        initLdap(connection.get(), initialDataLdif);
                    } catch (Exception e) {
                        throw new IllegalStateException(e);
                    } finally {
                        connection.get().close();
                    }
                } else {
                    throw new IllegalStateException("Cannot connect");
                }
            }

            private LDAPConnection createConnection() {
                LDAPConnectionOptions options = new LDAPConnectionOptions();
                options.setConnectTimeoutMillis((int) LDAP_CONNECT_TIMEOUT.toMillis());
                return new LDAPConnection(options);
            }

            private Optional<LDAPConnection> tryConnect(String address, Integer port) {
                LDAPConnection connection = createConnection();
                do {
                    try {
                        connection.connect(address, port);
                        Thread.sleep(WAIT_BETWEEN_RETRIES.toMillis());
                    } catch (Exception ignored) {}
                } while(!connection.isConnected());
                return Optional.of(connection);
            }

            private void initLdap(LDAPConnection connection, File initialDataLdif) throws Exception {
                LDAPConnection bindedConnection = bind(connection);
                LDIFReader r = new LDIFReader(initialDataLdif.getAbsoluteFile());
                Entry readEntry;
                while ((readEntry = r.readEntry()) != null) {
                    bindedConnection.add(new AddRequest(readEntry.toLDIF()));
                }
            }

            private LDAPConnection bind(LDAPConnection connection) throws Exception {
                UnboundidLdapClient.BindDnPassword bindDNAndPassword = getBindDNAndPassword();
                BindResult bindResult = connection.bind(bindDNAndPassword.getDn(), bindDNAndPassword.getPassword());
                if (!ResultCode.SUCCESS.equals(bindResult.getResultCode())) {
                    throw new Exception("Cannot init LDAP due to bind problem");
                }
                return connection;
            }
        };
    }
}

File: core/src/test/scala/tech/beshu/ror/utils/LdapContainer.scala
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.utils

import java.io.InputStream

import com.dimafeng.testcontainers.GenericContainer
import com.typesafe.scalalogging.StrictLogging
import com.unboundid.ldap.sdk.{AddRequest, LDAPConnection, ResultCode}
import com.unboundid.ldif.LDIFReader
import monix.eval.Task
import monix.execution.Scheduler.Implicits.global
import org.apache.logging.log4j.scala.Logging
import org.testcontainers.containers.wait.strategy.AbstractWaitStrategy
import tech.beshu.ror.acl.utils.ScalaOps.retryBackoff
import tech.beshu.ror.utils.LdapContainer.defaults

import scala.concurrent.duration._
import scala.language.postfixOps

class LdapContainer(name: String, ldapInitScript: String)
  extends GenericContainer(
    dockerImage = "osixia/openldap:1.1.7",
    env = Map(
      "LDAP_ORGANISATION" -> defaults.ldap.organisation,
      "LDAP_DOMAIN" -> defaults.ldap.domain,
      "LDAP_ADMIN_PASSWORD" -> defaults.ldap.adminPassword,
      "LDAP_TLS_VERIFY_CLIENT" -> "try"
    ),
    exposedPorts = Seq(defaults.ldap.port),
    waitStrategy = Some(new LdapWaitStrategy(name, ldapInitScript))
  ) {

  def ldapPort: Int = this.mappedPort(defaults.ldap.port)

  def ldapSSLPort: Int = this.mappedPort(defaults.ldap.sslPort)

  def ldapHost: String = this.containerIpAddress

  def stop(): Unit = {
    this.container.stop()
  }
}

object LdapContainer extends StrictLogging {

  object defaults {
    val connectionTimeout: FiniteDuration = 5 seconds
    val containerStartupTimeout: FiniteDuration = 5 minutes

    object ldap {
      val port = 389
      val sslPort = 636
      val domain = "example.com"
      val organisation = "example"
      val adminName = "admin"
      val adminPassword = "password"
      val bindDn: Option[String] = {
        Option(
          defaults.ldap.domain
            .split("\\.").toList
            .map(part => s"dc=$part")
            .mkString(","))
          .filter(_.trim.nonEmpty)
          .map(dc => s"cn=${defaults.ldap.adminName},$dc")
      }
    }

  }

}

private class LdapWaitStrategy(name: String,
                               ldapInitScript: String)
  extends AbstractWaitStrategy
    with Logging {

  override def waitUntilReady(): Unit = {
    logger.info(s"Waiting for LDAP container '$name' ...")
    Task(getClass.getResourceAsStream(ldapInitScript))
      .bracket(stream =>
        retryBackoff(ldapInitiate(stream), 15, 1 second, 1)
      )(stream =>
        Task(stream.close())
      )
      .onErrorHandle { ex =>
        logger.error("LDAP container startup failed", ex)
        throw ex
      }
      .runSyncUnsafe(defaults.containerStartupTimeout)
    logger.info(s"LDAP container '$name' stated")
  }

  private def ldapInitiate(ldapInitScriptInputStream: InputStream) = {
    runOnBindedLdapConnection { connection =>
      initLdapFromFile(connection, ldapInitScriptInputStream)
    }
  }

  private def initLdapFromFile(connection: LDAPConnection, scriptInputStream: InputStream) = {
    val reader = new LDIFReader(scriptInputStream)
    val entries = Iterator
      .continually(Option(reader.readEntry()))
      .takeWhile(_.isDefined)
      .flatten
      .toList
    Task
      .sequence {
        entries.map { entry =>
          Task(connection.add(new AddRequest(entry.toLDIF: _*)))
            .flatMap {
              case result if result.getResultCode == ResultCode.SUCCESS =>
                Task.now(())
              case result =>
                Task.raiseError(new IllegalStateException(s"Adding entry failed, due to: ${result.getResultCode}"))
            }
        }
      }
      .map(_ => ())
  }

  private def runOnBindedLdapConnection(action: LDAPConnection => Task[Unit]): Task[Unit] = {
    defaults.ldap.bindDn match {
      case Some(bindDn) =>
        Task(new LDAPConnection(waitStrategyTarget.getContainerIpAddress, waitStrategyTarget.getMappedPort(defaults.ldap.port)))
          .bracket(connection =>
            Task(connection.bind(bindDn, defaults.ldap.adminPassword))
              .flatMap {
                case result if result.getResultCode == ResultCode.SUCCESS => action(connection)
                case result => Task.raiseError(new IllegalStateException(s"LDAP '$name' bind problem - error ${result.getResultCode.intValue()}"))
              }
          )(connection =>
            Task(connection.close())
          )
      case None =>
        Task.raiseError(new IllegalStateException(s"Cannot create bind DN from LDAP config data"))
    }
  }
}
